jerry-core/ecma/base/ecma-helpers-value.c

```diff
@@ -37,6 +37,7 @@ JERRY_STATIC_ASSERT (sizeof (jmem_cpointer_t) <= sizeof (ecma_value_t),
 
 #ifdef ECMA_VALUE_CAN_STORE_UINTPTR_VALUE_DIRECTLY
 
+/* cppcheck-suppress zerodiv */
```

Comment:
Some comments for this suppression. It's definitely false warning, suppressing is the only way to fix this issue.
Cppcheck assumes the types and type sizes are fixed ("By default the platform which was used to compile Cppcheck is used."), but additionally it checks all combinations of if/ifdef macros. Unfortunately it is an invalid assumption, because the true case is for x86, the false case is for x86_64. It's incorrect to check the true case with x86_64 type sizes, which makes the static assert fails ( which is 1 / 0 == division by zero)

Here is a small example to show the same false warning:
```
#include <stdint.h>
#if UINTPTR_MAX <= UINT32_MAX
/* CPU: x86 / ARM */
enum { akarmi = 1 / (!!(sizeof (uintptr_t) <= sizeof (uint32_t)  ))};
#else
/* CPU: x86_64*/
enum { akarmi = 1 / (!!(sizeof (uintptr_t) > sizeof (uint32_t)  ))};
#endif
```

Reply:
Something is still itching me. There are other static asserts in the code that are mutually exclusive. E.g., in jerry-core/ecma/base/ecma-helpers-number.c:
```c
#if !ENABLED (JERRY_NUMBER_TYPE_FLOAT64)
JERRY_STATIC_ASSERT (sizeof (ecma_number_t) == sizeof (uint32_t),
                     size_of_ecma_number_t_must_be_equal_to_4_bytes);
#elif ENABLED (JERRY_NUMBER_TYPE_FLOAT64)
JERRY_STATIC_ASSERT (sizeof (ecma_number_t) == sizeof (uint64_t),
                     size_of_ecma_number_t_must_be_equal_to_8_bytes);
```
These don't trigger cppcheck warnings, and not because they are suppressed. What's the difference? Cannot we tweak the problematic code to get rid of the warning?

Reply:
BTW, in this code:
```c
#ifdef ECMA_VALUE_CAN_STORE_UINTPTR_VALUE_DIRECTLY

JERRY_STATIC_ASSERT (sizeof (uintptr_t) <= sizeof (ecma_value_t),
                     uintptr_t_must_fit_in_ecma_value_t);

#else /* !ECMA_VALUE_CAN_STORE_UINTPTR_VALUE_DIRECTLY */

JERRY_STATIC_ASSERT (sizeof (uintptr_t) > sizeof (ecma_value_t),
                     uintptr_t_must_not_fit_in_ecma_value_t);

#endif /* ECMA_VALUE_CAN_STORE_UINTPTR_VALUE_DIRECTLY */
```
what is the else branch trying to express? "uintptr_t must not fit in ecma_value_t"? must not? What will make the program misbehave? (The if branch is clear: if uintptr_t doesn't fit in ecma_value_t, quite some code will not work in the ECMA_VALUE_CAN_STORE_UINTPTR_VALUE_DIRECTLY-guarded code paths. But it doesn't seem to be the case the other way around.)

Reply:
> These don't trigger cppcheck warnings, and not because they are suppressed. What's the difference? Cannot we tweak the problematic code to get rid of the warning?

They are different. Both assert in the cited JERRY_NUMBER_TYPE_FLOAT64 example is true and aren't architecture (type size) dependent, the user can set JERRY_NUMBER_TYPE_FLOAT64.

The original issue is different, because the #if condition is a constant architecture dependent value. On x86 we get: #if true, on x86_64 we get #if false. The problem is that cppcheck shouldn't check the other assertion which will never be reached on the host architecture.